src/lib/ppx_register_event/register_event.ml

```diff
@@ -0,0 +1,197 @@
+open Core_kernel
+open Ppxlib
+
+let deriver = "register_event"
+
+(* module in Structured_events library *)
+let module_name = "Structured_events"
+
+let hash name =
+  let open Digestif.SHA256 in
+  let ctx0 = init () in
+  let ctx1 = feed_string ctx0 name in
+  get ctx1 |> to_raw_string
+
+let check_interpolations ~loc msg label_names =
+  match msg.pexp_desc with
+  | Pexp_constant (Pconst_string (s, _)) ->
+      (* check that every interpolation point $foo in msg has a matching label;
+       OK to have extra labels not mentioned in message
+    *)
+      let len = String.length s in
+      let interpolates =
+        let re = Str.regexp "\\$[a-z]\\([a-z]\\|[0-9]\\)*" in
+        let rec loop start acc =
+          if start >= len then acc
+          else
+            try
+              let offs = Str.search_forward re s start in
+              let s = Str.matched_string s in
+              let s_len = String.length s in
+              loop (offs + s_len) (String.sub s ~pos:1 ~len:(s_len - 1) :: acc)
+            with _ ->
+              (* wild-card match; actual match is deprecated Not_found *)
+              acc
+        in
+        loop 0 []
+      in
+      List.iter interpolates ~f:(fun interp ->
+          if not (List.mem label_names interp ~equal:String.equal) then
+            Location.raise_errorf ~loc
+              (Scanf.format_from_string
+                 (sprintf
+                    "The msg contains interpolation point \"$%s\" which is \
+                     not a field in the record"
+                    interp)
+                 "") )
+  | _ ->
+      ()
+
+let type_ext_str ~options ~path (ty_ext : type_extension) =
+  let add_module_qualifier s = String.concat ~sep:"." [module_name; s] in
+  let ctor, ctor_args, label_decls =
+    match ty_ext.ptyext_constructors with
+    (* record argument *)
+    | [ { pext_name
+        ; pext_kind= Pext_decl ((Pcstr_record labels as record), None)
+        ; _ } ] ->
+        (pext_name.txt, record, labels)
+    (* no arguments *)
+    | [{pext_name; pext_kind= Pext_decl ((Pcstr_tuple [] as tuple), None); _}]
+      ->
+        (pext_name.txt, tuple, [])
+    | _ ->
+        Location.raise_errorf ~loc:ty_ext.ptyext_path.loc
+          "Constructor in type extension must take a single record argument, \
+           or no argument"
+  in
+  let label_names =
+    List.map label_decls ~f:(fun {pld_name= {txt; _}; _} -> txt)
+  in
+  let has_record_arg = not @@ List.is_empty label_names in
+  let (module Ast_builder) = Ast_builder.make ty_ext.ptyext_path.loc in
```

Comment:
Missed this before, but this choice of location seems weird.. Doesn't this mean that our AST will have the same location as the `Structured_events.t` in the type extension?

Reply:
we could use the location of the `register_event` attribute ?